### DISCUSSION TITLE
Partial grouping of GET parameters

### DISCUSSION THREAD

**Discussion by Sanchoyzer** [fastapi/fastapi#12276] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
### main.py

from typing import Annotated, Literal

from fastapi import FastAPI, Depends, Query

from pydantic import BaseModel


class BaseParams(BaseModel):
    pass


class LimitOffsetMixin:
    limit: int = 100
    offset: int = 0


class OrderMixin:
    order_by: Literal["created_at", "updated_at"] = "created_at"


class BookFilters(BaseParams):
    title: str
    tags: list[str] | None = None


class BooksParams(LimitOffsetMixin, OrderMixin, BaseParams):
    filters: Annotated[BookFilters, Depends()]


class AuthorFilters(BaseParams):
    name: str
    countries: list[str] | None = None


class AuthorParams(LimitOffsetMixin, OrderMixin, AuthorFilters):
    pass


app = FastAPI()


@app.get("/books")
async def books(params: Annotated[BooksParams, Depends()]):
    assert params.filters.title
    assert params.filters.tags
    return {"ok": True}


@app.get("/authors")
async def authors(params: Annotated[AuthorParams, Query()]):
    filters = AuthorFilters(name=params.name, countries=params.countries)
    assert filters.name
    assert filters.countries
    return {"ok": True}


### test_1.py

from starlette.testclient import TestClient

from main import app


client = TestClient(app)


def test_books():
    r = client.get("/books", params={"title": "a", "tags": ["q", "w"]})
    assert r.status_code == 200, r.json()


def test_authors():
    r = client.get("/authors", params={"name": "a", "countries": ["q", "w"]})
    assert r.status_code == 200, r.json()
```

### Description

I use `Depends` to create separate GET parameters. For example:
```python
class BaseParams(BaseModel):
    pass

class LimitOffsetMixin:
    limit: int = 100
    offset: int = 0

class OrderMixin:
    order_by: Literal["created_at", "updated_at"] = "created_at"

class BookFilters(BaseParams):
    title: str
    # many other parameters, they are relative to the Book entity

class BooksParams(LimitOffsetMixin, OrderMixin, BaseParams):
    filters: Annotated[BookFilters, Depends()]

@app.get("/books")
async def books(params: Annotated[BooksParams, Depends()]):
    assert params.filters.title
```
And the same way I can use for creating more handlers. I just need to "overload" only the specific fields:
```python
class AuthorFilters(BaseParams):
    name: str
    # many other parameters, they are relative to the Author entity

class AuthorParams(LimitOffsetMixin, OrderMixin, BaseParams):
    filters: Annotated[AuthorFilters, Depends()]

@app.get("/authors")
async def authors(params: Annotated[AuthorParams, Depends()]):
    assert params.filters.name
```
The main benefit for me is that I can manage all "entity specific" parameters via one variable -- `params.filters`

And it works fine, but I've got a problem. When I try to use something more complex (e.g. `list[str]`), this system breaks
```python
class BookFilters(BaseParams):
    title: str
    tags: list[str] | None = None  # new field

@app.get("/books")
async def books(params: Annotated[BooksParams, Depends()]):
    assert params.filters.title
    assert params.filters.tags  # new check -- it'll be failed
```

There is an alternative way: use `Query` instead of `Depends`:
```python
class AuthorFilters(BaseParams):
    name: str
    countries: list[str] | None = None

class AuthorParams(LimitOffsetMixin, OrderMixin, AuthorFilters):
    pass

@app.get("/authors")
async def authors(params: Annotated[AuthorParams, Query()]):
    filters = AuthorFilters(name=params.name, countries=params.countries)  # additional manual grouping
    assert filters.name
    assert filters.countries
```
It works good, but I should group "entity specific" parameters manually

Is it possible to use hierarchy for grouping GET parameters and not lose support for more complex types (like `list[str]`)?

### Operating System

Linux

### Operating System Details

Ubuntu 24.4

### FastAPI Version

0.115.0

### Pydantic Version

2.9.2

### Python Version

3.12.2

**Comment by Kfir-G** [comment]
The issue you're running into happens because `Depends` works great for most parameters, but it struggles with more complex types like `list[str]` in GET requests. When you use `Depends` with nested models, FastAPI has a hard time parsing lists from query parameters.

### Quick Fixes:
1. **Use `Query()` for lists**: You can stick with `Depends` for most parameters, but switch to `Query()` when handling lists like `list[str]`.

   ```python
   from fastapi import Query
   class BookFilters(BaseParams):
       title: str
       tags: list[str] | None = Query(None)
   ```

2. **Manual Parsing**: Alternatively, manually handle the list parsing in your view while still using `Depends` for simpler parameters.

These solutions let you keep your structure without breaking list support in GET requests.

For more details, check out the [FastAPI docs on Query parameters](https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#import-query-and-annotated).

**Reply by Sanchoyzer** [reply]
@Kfir-G, thanks for your answer.

I've checked this code:
```python
...

class BookFilters(BaseParams):
    title: str
    tags: list[str] | None = Query(None)

class BooksParams(LimitOffsetMixin, OrderMixin, BaseParams):
    filters: Annotated[BookFilters, Depends()]

@app.get("/books")
async def books(params: Annotated[BooksParams, Depends()]):
    assert params.filters.tags
    return {"ok": True}

def test_books():
    r = client.get("/books", params={"title": "a", "tags": ["q", "w"]})
    assert r.status_code == 200, r.json()
```
and it seems that this quick fix doesn't work: `tags` are still empty

**Comment by yakubka** [comment]
FastAPI does not support grouping a subset of query parameters under a shared prefix in the URL or in the OpenAPI spec. All query parameters are listed flat in the spec regardless of how they are organized in the code.

The closest supported pattern is to use a Pydantic model as a query parameter container via `Depends`.

```python
from pydantic import BaseModel
from fastapi import Depends

class PaginationParams(BaseModel):
    page: int = 1
    size: int = 20

class FilterParams(BaseModel):
    status: str | None = None
    tag: str | None = None

@app.get("/items")
async def list_items(
    pagination: PaginationParams = Depends(),
    filters: FilterParams = Depends(),
):
    ...
```

This groups the parameters conceptually in code, and FastAPI will inject them as separate objects. The OpenAPI spec still lists all fields flat. If you need the spec to show them in groups, the only option is to use `openapi_extra` to manually inject the parameter descriptions with custom grouping metadata, but standard Swagger UI does not render those groups visually.
